### PR TITLE
Darkstalkers Chronicle: Add specializations and speedhacks to get it kinda playable

### LINKED ISSUE #6647: Darkstalkers Chronicles problem - characters not visible on select screen
Hi I'm currently using version ppsspp-v0.9.9-34-gb7ac539-windows-amd64 of the emulator and when I enter any game mode, arcade for example, I cannot see characters on select screen. When I select anything in blind and the stage loads up there is nothing visible, no hp bars, no characters, only stage background. It also occurred in 0.9.8 various versions. I downloaded the latest version for testing, but the problem is still happening. Is it possible to fix it somehow? In other 2D beat'em ups like Street Fighter or Samurai Showdown everything works fine.

### PR DESCRIPTION
**Pull request by hrydgard**
Darkstalkers Chronicles is a game that does bad things. Specifically it uses the PSP CPU to draw on top of GPU rendered graphics, which PPSSPP's hardware rendering can't normally handle.

This pull request is:

 * A compat setting that forces software rendering on for this game
 * Optimizations to the software rasterizer, specifically for rectangular draws (which avoids triangle interpolation math).
 * A hack to identify a specific stretch blit in Darkstalkers and replace it with a hardware stretch at the end

Together, it's enough to get Darkstalkers playable at a good speed on PC and kinda playable-ish on high powered mobile devices. And looking sharper than on the real PSP due to avoiding the extra stretch to 480x272. Fixes #6647 .

~~NOTE: This currently doesn't work on OpenGL, but will soon.~~ Works fine on all backends now!

Plenty more can be done to optimize the software rasterizer for this game and in general.

### CODE REVIEW

**Comment by LunaMoo** [comment]
It would be good if this hack also grayed out UI options not available for software renderer, so people wouldn't waste time changing their settings/post process effects and then wonder why none of this stuff does anything:3. Either way nice work.

 I guess this does more than optimizing software rendering, intentional or not it also fixed frameskipping which wasn't really working for this game in the past for software renderer and that should make it playable at much wider range of devices.


 Personally I had a plan to create some functions/dividing the awfuly large function this game uses to make it easier to hack around in hardware renderer, but first I was annoyed by some users and when I got back at it, I couldn't save function hashes reliably, maybe for the better since this way works well and improvements to software renderer are always nice to see, with some effects working only there:).

**Comment by hrydgard** [comment]
Yeah, I think this way is strictly better than hooking into the game - the software rendering specializations are universal (although of course not all games might hit exactly these paths), and I've managed to keep the other hacks very isolated and thus unlikely to risk anything else. Plus, it's really neat to see the resolution adjust to the game instead of an awkward stretch to 480x272, and then stretch again - it looks great.

Yes, disabling the unused settings should be done. Additionally, I think I consider it a bug that post-processing doesn't work with the software renderer, it wouldn't be that hard to do, just a bit of plumbing. I'll file a separate issue for that.

**Comment by hrydgard** [comment]
Fixed this to work on OpenGL ES, and curiously, this runs quite a bit better on OpenGL than on Vulkan on my S8+. Since it spends nearly all effort doing software rendering, this has to be a texture upload speed difference or something like that... very strange.

**Comment by hrydgard** [comment]
Turns out that the lovely stretch hack doesn't work if the game shows a save/load dialog (or, any other drawing happening after it). One way to fix that specific case would be to turn off the stretch hack if a dialog is visible, and one way to fix it in general, but with a one-frame glitch, would be to turn off the stretch hack the next frame if a draw happens after the stretch in the current frame. Even better would be to replay the stretch if a draw happens after. 

All this though is getting really close to the edge of what I find acceptable to do for a single game .. but the stretch hack really improves the experience a lot. Grr.

**Comment by hrydgard** [comment]
This is admittedly less than pretty, but takes a long-unplayable game and brings it with pretty much perfect graphics (better than the PSP due to avoiding a stretch) to Android devices.

The save/load dialog fix is especially ugly - we need to go back to normal rendering when a dialog is up, and return to our special stretch mode after but then we also need to clear the buffer, otherwise the save dialog briefly shows up whenever the stretch doesn't happen, like between matches.

Still looking into disabling the irrelevant settings, and finding nicer solutions to the hacks described above.

**Comment by LunaMoo** [comment]
Not sure if last commit affects performance, as this was without it, but made a quick test on Redmi Note 7:
Screenshot_2019-10-26-23-10-39-168_org ppsspp ppssppluna
Without much action it goes to around 46fps, either way since this also fixes frameskipping, it's easily playable at constant 30fps on very mediocre non-gaming phone.

Some notes:
 "Press Start" menu is a bit more demanding, and video intro becomes laggy with frameskip despite running 60 fps otherwise, but meh it doesn't matter, this game doesn't have story or anything anyway, it's just a generic fighter.
 Normal non-wide mode cuts the fps to ~ 12 ~  about 1/3 of wide mode and has broken frameskip, so not recommended to use outside of modern PC's which could brute force through it even before optimizations:).

**Comment by i30817** [comment]
Maybe it would be better to hack the game itself rather than the emulator.

**Comment by LunaMoo** [comment]
Nah.
 This game ~ weirdly positive reviews aside ~ was never popular on PSP - a platform which had many better fighting games with actual story, multiplayer and graphics made for PSP or at least scaled down from more powerful consoles. Darkstalkers - a game without a story will not be appealing to fanboys with multiplayer removed and a stretched blurry graphics when they could just get the original dreamcast version, that didn't suffered from neither of that and certainly had less input lag as well since all the shoddy PSP ports of old Capcom are infamous for noticeable input lag even on real hardware.

 If nobody cared to reprogram how this game renders it's graphics before this PR, it will certainly not change now after Henrik not only worked on performance, but also improved how the game looks, restoring Dreamcast quality at least to wide-screen mode:].

**Comment by i30817** [comment]
I was more thinking that 'no one cares' because no one has the skills. If hrydgard has the skills or can point the correct code locations out to another person, that might as well happen instead of a emulator hack.

**Comment by LunaMoo** [comment]
If you need info from someone else, you can't do shit as it takes a few minutes with debugging tools to understand everything.
 There's no info from here that could help with reprogramming the game itself, it's software rendering optimization and a hack which replaces a software stretch with a hardware one in specific conditions. It doesn't do anything to the bothersome graphics that would have to be done differently, it leaves it all to software renderer where it just works.

**Comment by hrydgard** [comment]
@LunaMoo  you might wanna test again, the last commit actually has a substantial effect on speed.

Still, yeah, avoid non-wide mode for now. I may do a specialized fix for that later, but for now, meh.

This is safe so I'm gonna merge this as-is, and clean up the presentation workarounds a little bit later.

**Comment by LunaMoo** [comment]
Yeah it did, on my phone fps rises above 50 fps frequently now and never drops under 40 during fight at least.

 A pity those optimizations don't do much to more complex games / 3D graphics, in Macross or Misshitsu No Sacrifice where mipmap effects only works in software I'm not sure if it even gained a single fps outside of some simple screens. Either way still nice to see work on software renderer:).

**Comment by hrydgard** [comment]
Hm, my 2 1/2 year old S8+ consistently does 60fps now, I guess it's faster than I give it credit for in raw CPU at least. But yeah, that's still quite an improvement.

**Comment by unknownbrackets** [comment]
Haven't really had time to look at this.  I had started on some of a softgpu jit a little while ago, but if the plan it to start adding game specific hacks I'm going to abandon that since obviously it would just be an unfun way to break all the hacks.

-[Unknown]

**Comment by hrydgard** [comment]
@unknownbrackets  It should be noted that the only actual hack here is just the presentation of the low resolution game buffer, where we bypass a stretch that ruins the image quality. This is clearly gated by a compat ini flag. The rest is just specialization of a specific couple of sets of state (rectangular textured draws with a specific blend mode etc), making it run way faster.

**Comment by hrydgard** [comment]
Also this game is a bit of a special case due to CPU drawing on top of GPU polygons - this way seemed like the least of all evils, and I have no plans to go further in this direction. If a jit is added, this will stay a little oddity in the non jit software rendering path, or be removed (though it's nice to support non jit devices).

**Comment by LunaMoo** [comment]
Small update on test from redmi note 7, because I never play games on my phone I didn't thought about such possibility, but probably my battery was low when I was testing last time, sorry for misleading results as I get 60fps now with battery at 90%.

 Heck even GOW runs pretty well between 55-60 with lots of action when previously when I tested it, I was getting like ~20fps.

**Comment by unknownbrackets** [comment]
Got it, well, in theory a jit path will give many of these speed boosts much more generically (though not the rectangle detection, although some games draw a series of rectangles with triangles.)  Also, I wonder if we want any threading of DrawSprite.

About battery, yeah.  That's why I added the battery warning stuff.  No one seems to expect that playing with low battery / automatic battery saver on would have any significant impact to performance.

-[Unknown]

**Comment by hrydgard** [comment]
DrawSprite could be threaded but in this game no draws except on the title screen is bigger than 32x32 so the way the triangle threading works, that's probably barely worth it due to sync overhead, and even less so given that the per-pixel cost is much lower in drawsprite.

File: assets/compat.ini [hunk 1]
```diff
@@ -683,3 +683,13 @@ UCET00844 = true
...
 UCED00971 = true
 UCUS98713 = true
+
+[ForceSoftwareRenderer]
+# Darkstalkers
+ULES00016 = true
+ULUS10005 = true
+
+[DarkStalkersPresentHack]
+# Darkstalkers
```
Comment (unknownbrackets) [inline assets/compat.ini]: Apparently these lists are missing ULJM05005?

-[Unknown]

Comment (hrydgard) [reply assets/compat.ini]: Ah, guess I missed this one since it's called Vampire Chronicles in certain regions.

**Pull request merged**